### DISCUSSION TITLE
ロボット->AIの情報

### DISCUSSION THREAD

**Discussion by HansRobo** [ibis-ssl/crane#393] [General]
現状送信している情報は以下

- リターンカウンタ
- キッカー状態
- 温度（モーターごと？）
- エラー情報
- エラー情報に紐づく値
- モーター電流値
- ボール検出（現状ボールセンサーのみ）
- ロボット角度
- Visionとの角度ズレ
- オドメトリ（位置）
- オドメトリ（速度）
- マウスセンサ生値
- バッテリー電圧

参考：https://github.com/ibis-ssl/crane/blob/develop/crane_msgs/msg/control/RobotFeedback.msg

一応パケット内容はこれになっているが、現状では全部にちゃんとした値が取れているわけではない

現状、「とりあえずこれ送っとけばええやろ！！！」というノリで送っているものが多いが、ここではしっかりと様々な面から考慮して過不足ない情報にしたい

**Comment by HansRobo** [comment]
## デバッグ用の情報

**Reply by ChikuwaRobo** [reply]
floatをたくさん送りたい。

**Comment by HansRobo** [comment]
## 運用上あれば嬉しい情報

バッテリー電圧など

**Reply by ChikuwaRobo** [reply]
バッテリー残量･モーター温度･キッカー温度に応じて負荷の高いポジションから移動できるといいかも。

**Comment by HansRobo** [comment]
## SSL-Visionでは取れない情報

**Reply by HansRobo** [reply]
- ビジョン範囲外でのロボット自己位置
  - マウスセンサーを使ったローカル位置推定結果を返すことでビジョンロストからも自動復帰が可能
- ローカルカメラによるボール位置検出
  - ボールがロボット で隠れても安定してボールプレイスメントができる。
  -  「ビジョン範囲外でのロボット自己位置」と組み合わせることでビジョン範囲外でもボールプレイスメントが可能になる
  - ずっと動かすと消費電力的に良くないのでボールのvisibilityが下がったときなど必要な場面で処理の起動をAI側から要求できると良いかも
- ボールセンサーによるボール接触情報
  - Visionだとせいぜい「ボールとキッカーが近い」程度の判定しかできない
  - 現状でのある程度動いているはず

**Comment by HansRobo** [comment]
## SSL-Visionの遅延を評価するための情報

Visionとの角度ズレと角速度一定回転モードと組み合わせると以下でVisionの遅延を評価できる。

```
(Visionの遅延) = (Visionとの角度ズレ) / (設定角速度)
```

ただし、このときローカル角度フィルタリングのVision成分をゼロにする必要がある

**Comment by ChikuwaRobo** [comment]
今のパケット内容についてのメモ

## 通信タイミングについて
CM4→マイコンへのコマンドのパケットを受信し終わったら送信する。
つまり、送信と受信タイミングは同期している。但し、CM4→マイコンは、CM4側についているカメラのデータを送るために、CM4側で一定サイクルで送信しているため、**AI側からすると、非同期で一方的にロボットからデータが送られてくるように見える**。
黄色がマイコン→CM4
水色がCM4→マイコン
![PXL_20231115_140549387](https://github.com/ibis-ssl/crane/assets/13376091/ff6a1a4f-2550-4941-916c-f2f0982bd03b)


## データ量について
現状64byte
マイコン→CM4間の通信が安定しないのと、CM4→マイコンとマイコン→CM4方向の通信が同時に発生するとうまく動かないっぽいのでデータ量を抑え気味にしてある。
本当はデバッグ用に、もっと大量にデータを流したい。
SPI化したらもっと増やせるはず。

## データ内容
### リターンカウンタ
AIコマンドのカウンタをそのまま返信する。
以前はロボット側でカウントしていたが、ロボットがAIコマンドを受信できているかのチェックのために変更

### キッカー状態
キック後一定時間は蹴れないようにしてあるので、それのカウンタ。
充電電圧を見るのとかでもいい気はする。

### 温度（モーターごと？）
モーター、キッカーコイル、キッカー充電回路の温度。

### エラー関係
エラー情報に紐づく値

### モーター電流値
本当は1mA単位とかで送りたいが、送れるデータ量が少ないので異常検知用程度に妥協している

### ボール検出（現状ボールセンサーのみ）
プレースメントなどに必須と言ってもいいので、一番重要まである

### ロボット角度
現状何に使っているというわけではなかった気がする

### Visionとの角度ズレ
レイテンシチェック(手動)に使っている

### オドメトリ
オムニオドメトリの座標と速度。
本当は、座標はオムニのみ、マウスのみ、Visionベースの3通りと、速度はオムニとマウスの2通りあるので、もっと増やしたい

### 電圧
バッテリー、キッカー充電電圧

**Reply by ChikuwaRobo** [reply]
https://github.com/ibis-ssl/G474_Orion_main/blob/multiple-control-mode/Core/Src/ai_comm.c#L11
今のデータ(64byte目以降は例として適当に入れてる)
